File: dart/test/update_quality_test.dart
import 'dart:io';

import 'package:test/test.dart';
import 'package:gilded_rose/gilded_rose.dart';

/// One test per row in `test/cases.csv`.
///
/// Each row encodes an input item (`name`, `sellIn`, `quality`) and the
/// expected state after a single `GildedRose.updateQuality()` call. The CSV
/// is generated once from the original implementation (see the Java
/// `GenerateCases` tool) and treated as the frozen oracle during refactoring.
const _casesPath = 'test/cases.csv';

void main() {
  final lines = File(_casesPath).readAsLinesSync()
    ..removeWhere((l) => l.trim().isEmpty);
  // Skip header.
  final rows = lines.skip(1);

  for (final line in rows) {
    final parts = line.split(';');
    final name = parts[0];
    final sellIn = int.parse(parts[1]);
    final quality = int.parse(parts[2]);
    final expectedSellIn = int.parse(parts[3]);
    final expectedQuality = int.parse(parts[4]);

    test(
      '$name (sellIn=$sellIn, quality=$quality) -> '
      'sellIn=$expectedSellIn, quality=$expectedQuality',
      () {
        // given an item built from the CSV row
        final item = Item(name, sellIn, quality);

        // when one day passes
        GildedRose(<Item>[item]).updateQuality();

        // then the resulting state matches the captured oracle
        expect(item.sellIn, expectedSellIn, reason: 'sellIn');
        expect(item.quality, expectedQuality, reason: 'quality');
      },
    );
  }
}

File: Java/src/test/java/com/gildedrose/UpdateQualityTest.java
package com.gildedrose;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvFileSource;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * One parameterized test per row in {@code src/test/resources/cases.csv}.
 *
 * Each row encodes an input item ({@code name}, {@code sellIn},
 * {@code quality}) and the expected state after a single
 * {@link GildedRose#updateQuality()} call. The CSV is generated once from
 * the original implementation by {@link GenerateCases} and treated as the
 * frozen oracle during refactoring.
 */
class UpdateQualityTest {

    @ParameterizedTest(name = "[{index}] {0} (sellIn={1}, quality={2}) -> sellIn={3}, quality={4}")
    @CsvFileSource(
            resources = "/cases.csv",
            numLinesToSkip = 1,
            delimiter = ';',
            quoteCharacter = '"'
    )
    void updateQuality_matchesExpectedBehaviour(
            String name,
            int sellIn,
            int quality,
            int expectedSellIn,
            int expectedQuality
    ) {
        // given an item built from the CSV row
        Item item = new Item(name, sellIn, quality);

        // when one day passes
        new GildedRose(List.of(item)).updateQuality();

        // then the resulting state matches the captured oracle
        assertThat(item.sellIn).as("sellIn").isEqualTo(expectedSellIn);
        assertThat(item.quality).as("quality").isEqualTo(expectedQuality);
    }
}
